src/litserve/specs/openai.py

```diff
@@ -74,6 +108,8 @@ class ChatCompletionRequest(BaseModel):
     presence_penalty: Optional[float] = 0.0
     frequency_penalty: Optional[float] = 0.0
     user: Optional[str] = None
+    tools: Optional[List[Tool]] = None
```

Comment:
this should be `tool_calls`. I sent the sample request to openai with streaming and get the following JSON response. 


```
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":"assistant","tool_calls":null},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":"call_kz3cshGHfs03xZAbjpRwyDTE","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"n\": \"S","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"an F","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"ranci","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"sco, C","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"A\", ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"\"unit","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"\": \"ce","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"lsiu","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":0,"id":null,"function":{"arguments":"s\"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":"call_jmYzksaRe2oms7Jp3Vuj3EyK","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"n\": \"T","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"okyo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"\", \"u","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"nit\": ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"\"cel","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"sius\"","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":1,"id":null,"function":{"arguments":"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":"call_9VH5q6ssHQuiZ41W1vmZqU2b","function":{"arguments":"","name":"get_current_weather"},"type":"function"}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"{\"lo","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"catio","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"n\": \"P","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"aris","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"\", \"u","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"nit\": ","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"\"cel","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"sius\"","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":[{"index":2,"id":null,"function":{"arguments":"}","name":null},"type":null}]},"finish_reason":null,"index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
{"id":"chatcmpl-9UX2xOCjsHM088WCt90mXonPUOSos","choices":[{"delta":{"content":null,"function_call":null,"role":null,"tool_calls":null},"finish_reason":"tool_calls","index":0,"logprobs":null}],"created":1717063899,"model":"gpt-4o-2024-05-13","object":"chat.completion.chunk","system_fingerprint":"fp_319be4768e","usage":null}
```

Reply:
Thank you @aniketmaurya for the sample response. It was really helpful.

BTW, In the above suggestions, `tools` is for ChatCompletionRequest, to receive available tools input.
I  have already added `tool_calls` in the chat message.
```python
class ChatMessage(BaseModel):
    role: str
    content: Union[str, List[Union[TextContent, ImageContent]]]
    name: Optional[str] = None
    tool_calls: Optional[List[ToolCall]] = None
    tool_call_id: Optional[str] = None
```